runelite-api/src/main/java/net/runelite/api/FreezeInfo.java

```diff
@@ -0,0 +1,194 @@
+	 */
+	@Getter
+	@Setter
+	private int immune;
+
+	/**
+	 * the type of freeze applied. To know which sprite to display
+	 * null upon initialisation and when no freeze is activate.
+	 * This value will also be null whilst immune and if this type is set during immunity
+	 * it's for the queued freeze.
+	 * As a result this value should never be null when the State is either QUEUED or FROZEN
+	 */
+	@Getter
+	@Setter
+	private FreezeType type;
+
+	/**
+	 * position we are frozen at, if we move from this point the freeze is automatically cancelled and immunity applied
+	 */
+	@Getter
+	private WorldPoint position;
+
+	public FreezeInfo()
+	{
+		this.queuedFreezes = new ArrayList<>();
+		this.frozen = 0;
+		this.immune = 0;
+		this.type = null;
+		this.position = null;
+	}
+
+	/**
+	 * Create a freeze info object with a queued freeze
+	 * @param type The freeze which will be applied
+	 * @param hasPid whether the target has pid or not, npcs always have pid
+	 */
+	public void queueFreeze(FreezeType type, int ticks, boolean hasPid)
+	{
+		this.queuedFreezes.add(new QueuedFreeze(type, ticks, hasPid));
+	}
+
+	/**
+	 * decrement all remaining ticks by 1, called every tick by the freeze manager.
+	 * for internal use only.
+	 */
+	public void decrementAll()
+	{
+		this.immune -= 1;
+		this.frozen -= 1;
+
+		for (Iterator<QueuedFreeze> qfIt = queuedFreezes.iterator(); qfIt.hasNext();)
+		{
+			QueuedFreeze qf = qfIt.next();
+			qf.decrement();
+
+			if (qf.getTicksTillActive() < 0)
+			{
+				qfIt.remove();
+				continue;
+			}
+		}
+	}
+
+	public boolean isFrozen()
+	{
+		return this.frozen > 0;
+	}
+
+	public boolean isImmune()
+	{
+		return this.immune > 0;
+	}
+
+	public boolean hasQueued()
+	{
+		return this.queuedFreezes.size() > 0;
+	}
+
+	public void startFreeze(FreezeType t, int frozenTicks, WorldPoint position)
+	{
+		this.immune = 0;
+		this.frozen = frozenTicks;
+		this.type = t;
+		this.position = position;
+	}
+
+	public void startImmunity()
+	{
+		this.immune = 6;
+		this.frozen = 0;
+		this.type = null;
+		this.position = null;
+	}
+
+	public void resetFreeze()
+	{
+		this.immune = 0;
+		this.frozen = 0;
+		this.type = null;
+		this.position = null;
+	}
+
+	public class QueuedFreeze
+	{
+		@Getter
+		private FreezeType type;
+
+		@Getter
+		private int lengthTicks;
+
+		@Getter
+		private int ticksTillActive;
+
+		/**
+		 *
+		 * @param type which type of freeze was applied
+		 * @param ticks pray halfing is determined on cast and not activation so ticks must be passed at this point and not later on
+		 * @param hasPid whether the target has pid or not, npcs should always have pid.
+		 */
+		public QueuedFreeze(FreezeType type, int ticks, boolean hasPid)
+		{
+			this.type = type;
+			this.lengthTicks = ticks;
+			this.ticksTillActive = hasPid ? 2 : 1;
+		}
+
+		public void decrement()
+		{
+			this.ticksTillActive -= 1;
+		}
+
+		public boolean shouldTrigger()
+		{
+			return (this.getTicksTillActive() == 0) ? true : false;
```

Comment:
Maybe I've been bitten too many times, but I feel like == 0 is always a bug waiting to happen, or at least, gets even worse when a bug happens and `decrement` is called twice. I'd feel safer if it was `<= 0`.